### PR TITLE
drivers: serial: qmsi: use codegen for driver instantiation

### PR DESCRIPTION
**Pull request by SavinayDharmappa**
Replace device instances code by codegen solution.
    
Extend binding to allow DTS extraction for device info.
    
Remove defines from dts.fixup that are now taken from DTS.

#9163

### CODE REVIEW

**Comment by SavinayDharmappa** [comment]
sanity is failing because , in boards/arm/reel_board/reel_board.dts "nxp,mma8652fc" used as compitable string but corresponding string is not found in any yaml file

**Comment by SebastianBoe** [comment]
@SavinayDharmappa : Should this be marked as DNM until it's dependent PR's are merged?

File: doc/subsystems/codegen/principle.rst [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+.. contents::
+   :depth: 2
+   :local:
+   :backlinks: top
+
+Principle
+---------
+
+.. image:: codegen_principle.png
```
Comment (dbkinder) [inline doc/subsystems/codegen/principle.rst]: Seems like these four pictures could use some words to explain what they're showing...

**Review by dbkinder — APPROVED** [review]
seems odd to use this PR to add documentation of Codegen, but OK.

**Comment by nashif** [comment]
closing now until code-generation design/discussion is complete.

**Pull request closed without merging**